cloudflare/resource_cloudflare_spectrum_application_test.go

```diff
@@ -0,0 +1,287 @@
+					resource.TestCheckResourceAttr(name, "protocol", "tcp/22"),
+					resource.TestCheckResourceAttr(name, "origin_direct.#", "1"),
+					resource.TestCheckResourceAttr(name, "origin_direct.0", "tcp://120.120.102.10:23"),
+					resource.TestCheckResourceAttr(name, "origin_port", "22"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccCloudflareSpectrumApplication_OriginDNS(t *testing.T) {
+	var spectrumApp cloudflare.SpectrumApplication
+	zone := os.Getenv("CLOUDFLARE_DOMAIN")
+	rnd := acctest.RandString(10)
+	name := "cloudflare_spectrum_application." + rnd
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudflareSpectrumApplicationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareSpectrumApplicationConfigOriginDNS(zone, rnd),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareSpectrumApplicationExists(name, &spectrumApp),
+					testAccCheckCloudflareSpectrumApplicationIDIsValid(name),
+					resource.TestCheckResourceAttr(name, "protocol", "tcp/22"),
+					resource.TestCheckResourceAttr(name, "origin_dns.#", "1"),
+					resource.TestCheckResourceAttr(name, "origin_dns.0.name", fmt.Sprintf("ssh.origin.%s", zone)),
+					resource.TestCheckResourceAttr(name, "origin_port", "22"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccCloudflareSpectrumApplication_Update(t *testing.T) {
+	var spectrumApp cloudflare.SpectrumApplication
+	var initialId string
+	zone := os.Getenv("CLOUDFLARE_DOMAIN")
+	rnd := acctest.RandString(10)
+	name := "cloudflare_spectrum_application." + rnd
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckCloudflareSpectrumApplicationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareSpectrumApplicationConfigBasic(zone, rnd),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareSpectrumApplicationExists(name, &spectrumApp),
+					testAccCheckCloudflareSpectrumApplicationIDIsValid(name),
+					resource.TestCheckResourceAttr(name, "origin_direct.0", "tcp://120.120.102.10:23"),
+				),
+			},
+			{
+				PreConfig: func() {
+					initialId = spectrumApp.ID
+				},
+				Config: testAccCheckCloudflareSpectrumApplicationConfigBasicUpdated(zone, rnd),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflareSpectrumApplicationExists(name, &spectrumApp),
+					testAccCheckCloudflareSpectrumApplicationIDIsValid(name),
+					func(state *terraform.State) error {
+						if initialId != spectrumApp.ID {
+							// want in place update
+							return fmt.Errorf("spectrum application id is different after second config applied ( %s -> %s )",
+								initialId, spectrumApp.ID)
+						}
+						return nil
+					},
+					resource.TestCheckResourceAttr(name, "origin_direct.0", "tcp://81.120.102.10:23"),
```

Comment:
We should use reserved IP here.
```suggestion
					resource.TestCheckResourceAttr(name, "origin_direct.0", "tcp://192.0.2.2:23"),
```